### PR TITLE
Improve naming consistency and semantics

### LINKED ISSUE #99: Improve naming consistency and semantics
I think it could bring a lot added value compared to Preview1 (and avoid carrying some technical debt) to have a more consistent and understandable API by somebody who has not a deep POSIX knowledge, so I am wondering if we could use the opportunity of Preview 2 and its big bang change nature to refine a bit some names that looks maybe a bit too POSIX-specific while keeping something reasonsably close to POSIX to keep easy semantic matching with WASI.

The goal of the proposed changes are:
- Make the API more consistent and predicatable
- Reduce the cognitive load involved by some inconsistencies
- Allow better APIs for language bindings generated automatically

The proposed changes below are:
 - Remove unneeded prefixes potentially not needed due to the namespaced nature of WIT interfaces
 - Expand meaningful prefix to their full name except for very long names ("synchronize" or "Synchronization" remain "sync")
 - Use `-` as a name separator consistently
 - Choose either singular (my preference) or plural consistently for enums

**Types**
- `o-flags` -> `open-flag`
   - `excl` -> `exclusive`
   - `trunc` -> `truncate`
- `errno` -> `error-code` : those are not numbers anymore and this new name is consistent with the documentation with better semantics
  - `deadlk` -> `deadlock`
  - `dquot` -> `quota`
  - `fbig` -> `file-too-large`
  - `ilseq` -> `Illegal-byte-sequence`
  - `inprogress` -> `in-progress`
  - `intr` -> `interrupted`
  - `inval` -> `invalid`
  - `isdir` -> `is-directory`
  - `mlink`-> `too-many-links`
  - `msgsize` -> `message-size`
  - `nametoolong` -> `name-too-long`
  - `no-dev` -> `no-device`
  - `noent` -> `no-entry`
  - `nolck` -> `no-lock`
  - `nomem` -> `memory`
  - `nospc` -> `space`
  - `nosys` and potentially `notsup` -> `unsupported`
  - `notdir` -> `not-directory`
  - `notempty` -> `not-empty`
  - `notrecoverable` -> `not-recoverable`
  - `notty` -> `no-tty`
  - `nxio` -> `io`
  - `perm` -> `not-permitted`
  - `rofs` -> `read-only`
  - `spipe` -> `invalid-seek`
  - `txtbsy` -> `busy`
  - `xdev` -> `cross-device`
- `dir-entry-stream` -> `directory-entry-stream`
- `dir-entry` -> `directory-entry`
  -  `ino` -> `inode`
- `descriptor-stat`
  - `dev` -> `device`
  - `ino` -> `inode`
  - `linkcount` -> `link-count`
  - `atim`-> `data-access-timestamp`
  - `mtim` -> `data-modification-timestamp`
  - `ctim` -> `status-change-timestamp`
- `descriptor-flags` -> `descriptor-flag`
  - `nonblock` ->  `non-blocking`
  - `sync` -> `file-integrity-sync`
  - `sync` -> `data-integrity-sync`
  - `rsync` -> `requested-write-sync`
- `at-flags` -> `path-flag` : consistent use of singular with enums and maybe `path` is more meaningul than `at`?

**Functions**
- `fadvise` -> `advise`
- `datasync` - `sync-data`
- `flags` -> `get-flags`
- `type` -> `get-type`
- `pread` -> `read`
- `pwrite` -> `write`
- `readdir` -> `read-directory`
- `read-dir-entry` -> `read-directory-entry`
- `drop-dir-entry-stream` -> `drop-directory-entry-stream`

If accepted, those principles probably make sense for other WASI repositories and some common guidelines could be created, but let's first discuss the principle here since filesystem is probably the most complex one with the most historical baggage.

### PR DESCRIPTION
**Pull request by sdeleuze**
The goal of the proposed changes are:
- Make the API more consistent and predictable
- Reduce the cognitive load involved by some inconsistencies
- Allow better APIs for language bindings generated automatically

The proposed changes below are:
- Remove unneeded prefixes potentially not needed due to the namespaced nature of WIT interfaces
- Expand meaningful prefix to their full name except for very long names (`synchronize` or `Synchronization` remain `sync`)
- Use `-` as a name separator consistently
- Use singular for enums
- Use plural for flags

Closes #99

### CODE REVIEW

**Comment by sdeleuze** [comment]
@sunfishcode For now, I updated only `wit/wasi-filesystem.wit.md`. Not sure if and how I am expected to update `wasi-filesystem.md` and `wasi-filesystem.html` and if that process is manual or automated. Please let me know if I should do something about that.

File: wit/wasi-filesystem.wit.md [hunk 1]
```diff
@@ -113,54 +113,54 @@ flags descriptor-flags {
...
 }
 ```
 
-## `mode`
+## `modes`
 ```wit
 /// Permissions mode used by `open-at`, `change-file-permissions-at`, and
 /// similar.
-flags mode {
+flags modes {
```
Comment (sdeleuze) [inline wit/wasi-filesystem.wit.md:163]: Could be potentially `permissions`.

File: wit/wasi-filesystem.wit.md [hunk 2]
```diff
@@ -226,89 +226,87 @@ record dir-entry {
...
     /// No locks available.
-    nolck,
+    no-lock,
     /// Not enough space.
-    nomem,
+    insufficient-memory,
     /// No space left on device.
-    nospc,
-    /// Function not supported.
-    nosys,
```
Comment (sdeleuze) [inline wit/wasi-filesystem.wit.md:285]: In this PR, I remove `nosys` (Function not supported) in favor of the new `unsupported` (Not supported). Another option if we want to not change the granualrity of error codes is to keep `nosys` and rename it to `function-unsupported`, but having both `unsupported` and `function-unsupported` looked a bit confusing to me, hence my proposal to translate both `nosys` and `notsup` to `unsupported` (but no strong take on that since I am not a POSIX expert). I also removed the comment about socket since I think it is now out of the scope.

File: wit/wasi-filesystem.wit.md [hunk 3]
```diff
@@ -226,89 +226,87 @@ record dir-entry {
...
 /// Error codes returned by functions.
 /// Not all of these error codes are returned by the functions provided by this
 /// API; some are used in higher-level library layers, and others are provided
 /// merely for alignment with POSIX.
-enum errno {
+enum error-code {
     /// Permission denied.
     access,
     /// Resource unavailable, or operation would block.
     again,
```
Comment (sunfishcode) [inline wit/wasi-filesystem.wit.md:239]: What would you think about changing `again` to `would-block`? In POSIX, `EAGAIN` is permitted to be equal to `EWOULDBLOCK` and the errors are the same on many popular platforms. And `would-block` is more descriptive of what the code is for than `again`.

Comment (sdeleuze) [reply wit/wasi-filesystem.wit.md]: Indeed, `would-block` is much more descriptive, I have updated the PR accordingly.

**Comment by sdeleuze** [comment]
I did some last refinements to remove elements not present anymore like `append`, change old names references to new names, and use `text-file-busy` for the old `txtbsy` in order to avoid a duplicate entry with `busy`.

**Comment by sunfishcode** [comment]
This looks good to me. I'm going to take the liberty of merging this sooner rather than later, as it's a big patch that will create merge conflicts with other changes in flight. We can continue to iterate on the names here. We can also continue to disucss whether this whole direction of moving away from POSIX's spelling of all these identifiers, if anyone wishes to.

**Comment by codefromthecrypt** [comment]
I think this will make things harder to map to the syscall layer. Maybe we might want to go back and document what these relate to? for example, nxio is obvioulsy ENXIO, but the replacement takes a lot of thinking

**Comment by codefromthecrypt** [comment]
Case in point, this diff

```diff
    /// No such device or address.
-    nxio,
+    no-such-device,
```

Probably for all of them, we should do something to remember what they were, extending the docs perhaps.

```
-    /// No such device or address.
+    /// No such device or address, e.g. ENXIO
```

If there are going to me a lot of renames, the docs would at least be stable and instructive as mapping err numbers is one of the most important things to do correctly.

**Comment by sdeleuze** [comment]
Yeah, I will submit a complementary PR that document the matching with POSIX in comments.

**Comment by codefromthecrypt** [comment]
thanks, @sdeleuze!

**Comment by swankjesse** [comment]
This is nice standalone but I think it’s a mistake. It makes WASI special where it aught not be.

I can’t bring my existing understanding of ‘dsync’ etc. WASI.

I can’t directly use StackOverflow answers from UNIX programming environments to WASI.

I can’t as easily port programs to WASI.

Finally I don’t expect most developers will use these APIs directly; they’ll be underlying primitives delegated to by some higher level abstraction. For example, Kotlin will use its own multiplatform streams and paths abstractions. This makes building that more difficult because it’s now dissimilar from other platforms.

**Comment by sdeleuze** [comment]
@swankjesse I understand the concern. I think we have tried to stay pretty close to POSIX semantics in order to allow people knowing POSIX to reuse that knowledge and we I think we can keep the benefits of those changes by documentation explicitly the matches with POSIX where that makes sense.

Regardless of my changes, WASI Preview2+ is not an exact mapping to POSIX for good reasons (Preview1 feedback, socket dedicated support, Wasm Component model), and resources support will IMO brings even more differences. Pretending WASI is POSIX with implicit similar names can also be source of confusion.

That said, I should have been more careful on documenting the link between the 2, see [this related PR](https://github.com/WebAssembly/wasi-filesystem/pull/104) for `error-code` to `errno` matching. Looks better to me that the kind of implicit matching using "remove leading `E` and lowercase" there was before those changes on `errno`. I will probably submit more of those to finish documenting the matching with POSIX semantics where that make sense. I will likely provide other PR to clarify the matching with POSIX semantics.

If you think we have been too far on some changes that introduces unnecessary noise (`dsync`, `sync`, `fsync` maybe?), feel free to provide feedback in order to allow us to refine.

**Comment by swankjesse** [comment]
Awesome. Documenting the mapping is good stuff.

**Comment by codefromthecrypt** [comment]
FYI, when wazero did wasi in the first place, each function took a *lot* of research because the docs were very minimal and also there was no spec to refer to in lieu of docs. As this drifts from POSIX and desires to do that, I think more docs will be required here (since you can't say "just look at POSIX anymore")

Here's an example of one function docs in WASI, noting that the new wasi via the component model decouples itself from being an ABI due to the canonical ABI thing being in a different repo. IOTW, memory output in binary isn't something needed in the same docs anymore.

TL;DR; it might help especially to remember others have to go back and document things missing, and when they can't rely on links to posix stuff for clarity, it may lead to longer docs or the need for a companion spec.
```go
// clockResGet is the WASI function named "clock_res_get" that returns the
// resolution of time values returned by "clock_time_get".
//
// # Parameters
//
//   - id: clock ID to use
//   - resultResolution: offset to write the resolution to api.Memory
//   - the resolution is an uint64 little-endian encoding
//
// Result (Errno)
//
// The return value is ErrnoSuccess except the following error conditions:
//   - ErrnoNotsup: the clock ID is not supported.
//   - ErrnoInval: the clock ID is invalid.
//   - ErrnoFault: there is not enough memory to write results
//
// For example, if the resolution is 100ns, this function writes the below to
// api.Memory:
//
//	                                   uint64le
//	                   +-------------------------------------+
//	                   |                                     |
//	         []byte{?, 0x64, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ?}
//	resultResolution --^
//
// Note: This is similar to `clock_getres` in POSIX.
// See https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#-clock_res_getid-clockid---errno-timestamp
// See https://linux.die.net/man/3/clock_getres
```
